clippy_lints/src/xor_used_as_pow.rs

```diff
@@ -0,0 +1,79 @@
+use crate::utils::{span_lint_and_help, span_lint_and_sugg};
+use if_chain::if_chain;
+use rustc_ast::{BinOpKind, Expr, ExprKind, LitKind};
+use rustc_errors::Applicability;
+use rustc_lint::{EarlyContext, EarlyLintPass};
+use rustc_middle::lint::in_external_macro;
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for use of `^` operator when exponentiation was intended.
+    ///
+    /// **Why is this bad?** This is most probably a typo.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    ///
+    /// ```rust,ignore
+    /// // Bad
+    /// 2 ^ 16;
+    ///
+    /// // Good
+    /// 1 << 16;
+    /// 2i32.pow(16);
+    /// ```
+    pub XOR_USED_AS_POW,
+    correctness,
+    "use of `^` operator when exponentiation was intended"
+}
+
+declare_lint_pass!(XorUsedAsPow => [XOR_USED_AS_POW]);
+
+impl EarlyLintPass for XorUsedAsPow {
+    fn check_expr(&mut self, cx: &EarlyContext<'_>, expr: &Expr) {
+        if_chain! {
+            if !in_external_macro(cx.sess, expr.span);
+            if let ExprKind::Binary(op, left, right) = &expr.kind;
+            if BinOpKind::BitXor == op.node;
+            if let ExprKind::Lit(lit) = &left.kind;
+            if let LitKind::Int(lhs, _) = lit.kind;
+            if let ExprKind::Lit(lit) = &right.kind;
+            if let LitKind::Int(rhs, _) = lit.kind;
```

Comment:
When checking if `rhs` is negative, it seems like we can't simply do a check to see if `rhs < 0` since `rhs` is a `u128` with an associated `LitIntType`. If the literal in the source code is signed, should I interpret the `rhs` as a signed integer and then check? I'm unsure where stuff like this would be in the docs.

Reply:
If the RHS is negative, the binary operation should be `LIT ^ NEG(LIT)` instead of `LIT ^ LIT`.